File: Source/CSharp/Expressive/Expressive/Expressions/FunctionExpression.cs
using System;
using System.Collections.Generic;

namespace Expressive.Expressions
{
    internal class FunctionExpression : IExpression
    {
        private readonly Func<IExpression[], IDictionary<string, object>, object> _function;
        private readonly string _name;
        private readonly IExpression[] _parameters;

        internal FunctionExpression(string name, Func<IExpression[], IDictionary<string, object>, object> function, IExpression[] parameters)
        {
            _name = name;
            _function = function;
            _parameters = parameters;
        }

        #region IExpression Members

        public object Evaluate(IDictionary<string, object> variables)
        {
            return _function(_parameters, variables);
        }

        #endregion
    }
}

File: Source/Java/Expressive/src/main/java/com/bijington/expressive/expressions/FunctionExpression.java
package com.bijington.expressive.expressions;

import com.bijington.expressive.functions.IFunction;

import java.util.Map;

/**
 * Created by shaun on 30/06/2016.
 */
public class FunctionExpression implements IExpression {
    private final IFunction _function;
    private final String _name;
    private final IExpression[] _parameters;

    public FunctionExpression(String name, IFunction function, IExpression[] parameters)
    {
        _name = name;
        _function = function;
        _parameters = parameters;
    }

    @Override
    public Object evaluate(Map<String, Object> variables) {
        return _function.evaluate(_parameters);
    }
}
